plugins/inputs/amqp_consumer/amqp_consumer.go

```diff
@@ -1,144 +1,246 @@
+	return `
+  ## AMQP url
+  url = "amqp://localhost:5672/influxdb"
+  ## AMQP exchange
+  exchange = "telegraf"
+  ## Auth method. PLAIN and EXTERNAL are supported
+  # auth_method = "PLAIN"
+  ## Binding Key
+  binding_key = "#"
+
+  ## Maximum number of messages server should give to the worker.
+  prefetch = 50
+
+  ## Optional SSL Config
+  # ssl_ca = "/etc/telegraf/ca.pem"
+  # ssl_cert = "/etc/telegraf/cert.pem"
+  # ssl_key = "/etc/telegraf/key.pem"
+  ## Use SSL but skip chain & host verification
+  # insecure_skip_verify = false
+
+  ## Data format to output.
+  ## Each data format has it's own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
   data_format = "influx"
 `
 }
 
-// SetParser satisfies the telegraf.ServiceInput interface
-func (rmq *AMQPConsumer) SetParser(parser parsers.Parser) {
-	rmq.parser = parser
+func (a *AMQPConsumer) Description() string {
+	return "AMQP consumer plugin"
+}
+
+func (a *AMQPConsumer) SetParser(parser parsers.Parser) {
+	a.parser = parser
 }
 
-// Gather satisfies the telegraf.ServiceInput interface
 // All gathering is done in the Start function
-func (rmq *AMQPConsumer) Gather(_ telegraf.Accumulator) error {
+func (a *AMQPConsumer) Gather(_ telegraf.Accumulator) error {
 	return nil
 }
 
+func (a *AMQPConsumer) createConfig() (*amqp.Config, error) {
+	// make new tls config
+	tls, err := internal.GetTLSConfig(
+		a.SSLCert, a.SSLKey, a.SSLCA, a.InsecureSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse auth method
+	var sasl []amqp.Authentication // nil by default
+
+	if strings.ToUpper(a.AuthMethod) == "EXTERNAL" {
+		sasl = []amqp.Authentication{&externalAuth{}}
+	}
+
+	config := amqp.Config{
+		TLSClientConfig: tls,
+		SASL:            sasl, // if nil, it will be PLAIN
+	}
+	return &config, nil
+}
+
 // Start satisfies the telegraf.ServiceInput interface
-func (rmq *AMQPConsumer) Start(acc telegraf.Accumulator) error {
+func (a *AMQPConsumer) Start(acc telegraf.Accumulator) error {
+	amqpConf, err := a.createConfig()
+	if err != nil {
+		return err
+	}
 
-	// Create queue connection and assign it to AMQPConsumer
-	conn, err := amqp.Dial(fmt.Sprintf("amqp://%v:%v@%v:%v", rmq.Username, rmq.Password, rmq.AmqpHost, rmq.AmqpPort))
+	conn, err := amqp.DialConfig(a.URL, *amqpConf)
 	if err != nil {
-		return fmt.Errorf("%v: Failed to connect to AMQP", err)
+		return err
 	}
-	rmq.conn = conn
+	a.conn = conn
 
 	// Create channel and assign it to AMQPConsumer
 	ch, err := conn.Channel()
 	if err != nil {
 		return fmt.Errorf("%v: Failed to open a channel", err)
 	}
-	rmq.ch = ch
+	a.ch = ch
+
+	err = ch.ExchangeDeclare(
+		a.Exchange, // name
+		"topic",    // type
+		true,       // durable
+		false,      // auto-deleted
+		false,      // internal
+		false,      // no-wait
+		nil,        // arguments
+	)
+	if err != nil {
+		return fmt.Errorf("Failed to declare an exchange: %s", err)
```

Comment:
I usually use `%s` for errors, is there a benefit to using `%v`? if err != nil then the string format will always work properly, I believe it will just print whatever err.Error() returns